File: algorithm/101. Symmetric Tree/101.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return 1;
        return helper(root->left, root->right);
    }
    
    bool helper(TreeNode* left1, TreeNode * right1)
    {
        if(!left1 &&  !right1)    return true;
        else if(!left1 || !right1) return false;
        else if(left1->val != right1->val) return false;
        
        return helper(left1->left,right1->right)&&helper(left1->right,right1->left);
    }
};

File: algorithm/101. Symmetric Tree/101.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null)    return true;
        return helper(root.left, root.right);
    }
    
    
    private boolean helper(TreeNode leftnode, TreeNode rightnode)
    {
        if(leftnode == null && rightnode == null)   return true;
        else if(leftnode == null || rightnode == null)  return false;
        else if(leftnode.val != rightnode.val)  return false;
        return helper(leftnode.left, rightnode.right)&&helper(leftnode.right, rightnode.left);
    }
}
